Components: Fix DateTimePicker timezone handling for non-string values

aduth opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Fixes an issue with `DateTimePicker` after #73444 where `Date` or `number` or `string` (with timezone) values passed as `currentDate` would be treated inconsistently from timezoneless strings, potentially resulting in mismatches between displayed and changed values similar to what was seen previously in #67480.

Related to: #67480
Follow-up to #73444
Fixes #73981

## Why?

The intent of the changes with #73444 were to normalize internal handling of dates within the component to use UTC time, but the changes only applied to values passed as timezoneless strings. In current Gutenberg usage, this is primarily how the component is used, since this is how WordPress represents dates.

Since the component internally assumes UTC times, this behavior should also extend to values passed in other formats (date objects, timestamps).

## How?

Updates `inputToDate` to consistently coerce all input values to UTC time.

## Testing Instructions

Verify included tests pass:

- `npm run test:unit packages/components/src/date-time/date-time/test/index.tsx`
- `npm run test:unit packages/components/src/date-time/test/utils.test.ts`

I was able to identify one Gutenberg usage where this bug previously existed, in [the initialization of the "Date" block](https://github.com/WordPress/gutenberg/blob/90c242a3020dc2ed8865cff38e5a33d7aa318fbd/packages/block-library/src/post-date/edit.js#L77).

To verify:

1. Set your local system timezone to a different date from the site's timezone (see instructions in #73444)
2. Go to "Posts" > "Add Post"
3. Insert a "Date" block
4. In the block toolbar, click "Change Date"
5. Verify that the initial date is accurately displayed as the current time on the site

## Screenshots or screencast <!-- if applicable -->\

These screenshots show a before/after of the Date block instructions described above, where my local time (1:52am UTC+10.5 the next day) is incorrectly shown, compared to the site's actual current time (3:22pm UTC).

Before|After
---|---
<img width="392" height="631" alt="Screenshot 2025-12-11 at 1 52 31 AM" src="https://github.com/user-attachments/assets/95d8c97a-8058-4574-ae66-d261564784cb" />|<img width="302" height="590" alt="Screenshot 2025-12-11 at 1 52 54 AM" src="https://github.com/user-attachments/assets/e37ae7b2-7930-4853-acf9-e9687f71fe29" />

aduth committed:
Components: Fix DateTimePicker timezone handling for non-string values

aduth committed:
Ensure more effective regression testing by mismatch timezone

tyxla commented on packages/components/src/date-time/utils.ts:
```diff
@@ -27,9 +25,13 @@ export function inputToDate( input: Date | string | number ): Date {
 		// See: https://tc39.es/ecma262/#sec-date-time-string-format
 		const hasTimezone = /Z|[+-]\d{2}(:?\d{2})?$/.test( input );
-		return hasTimezone ? new Date( input ) : new UTCDateMini( input + 'Z' );
+		return new UTCDateMini( hasTimezone ? new Date( input ) : input + 'Z' );
```
Can this break any existing use cases? Do you feel like we're well covered by tests?

aduth replied:
These changes should un-break some of the assumptions introduced by changes in #73444 for values other than timezoneless strings, since the changes there start introducing assumptions around UTC times in how dates are displayed. Although curiously none of these changes have affected how those date changes are serialized, so I'd be curious to test if this has always been somewhat broken for these types of date values.

As far as tests are concerned, there's a lot of different factors like local vs. server timezones, date object shapes, daylight savings time, etc, but I'd like to think (especially after the changes here) we have exhaustive branch coverage. For example, we previously had no unit test coverage for `inputToDate` covering timestamps.

tyxla replied:
Maybe we can run some AI prompts to see if any additional coverage ideas come up?

aduth replied:
Here's what AI had to say:

>### Coverage Assessment
>
>| Test Category | Coverage | Notes |
>|---------------|----------|-------|
>| String inputs (timezoneless) | ✅ Excellent | Multiple timezone scenarios across unit + integration tests |
>| String inputs (with timezone) | ✅ Good | Z, +offset, -offset all covered |
>| Date object inputs | ✅ Good | Unit test + DateTimePicker integration test |
>| Timestamp inputs | ✅ Good | Unit test + DateTimePicker integration test |
>| Timezone edge cases | ✅ Excellent | DST transitions, timezone mismatches tested |
>| Component usage in DatePicker | ⚠️ Adequate | Tests use strings; no explicit Date/timestamp testing |
>| Component usage in TimePicker | ⚠️ Adequate | Tests use strings; no explicit Date/timestamp testing |
>
>### Potential Gaps
>
>1. **TimePicker with non-string inputs**: While `DateTimePicker` tests Date objects and timestamps, the standalone `TimePicker` tests only use string inputs. However, since `TimePicker` uses the same `inputToDate` function and it's tested at the unit level, this is likely acceptable.
>
>2. **DatePicker with non-string inputs**: Same as above — tests only use strings. Again, covered by unit tests.
>
>3. **Invalid input handling**: No explicit tests for malformed inputs (e.g., invalid date strings). The function doesn't have explicit error handling, so this may be intentional.
>
>### Verdict
>
>**The test coverage is quite exhaustive for protecting against regressions**, particularly because:
>
>1. The PR commits (`718c5334126` and `13dcb543d54`) specifically added comprehensive timezone handling tests
>2. The unit tests mock multiple timezones to catch day-shift bugs
>3. The `DateTimePicker` integration tests cover the most critical real-world scenarios with timezone mismatches and DST transitions
>4. All three input types (strings, Date objects, timestamps) are tested
>
>The tests are well-designed to catch timezone-related regressions, which appears to be the core concern of PR 73887's fix for "DateTimePicker timezone handling for non-string values."

tyxla commented on packages/components/src/date-time/utils.ts:
```diff
@@ -27,9 +25,13 @@ export function inputToDate( input: Date | string | number ): Date {
+	// Date objects and number timestamps represent specific UTC moments.
+	// Convert to milliseconds since epoch for consistent UTC handling.
+	const time = input instanceof Date ? input.getTime() : input;
+	return new UTCDateMini( time );
```
Same question as above

tyxla reviewed (approved):
Thanks @aduth, let's get this in 👍

aduth committed:
Components: Add DateTime test coverage for treating input dates as configured offset

aduth committed:
Date: Fix handling date utilities when passing timezone as 0

`0` is a valid UTC offset, but the previous logic incorrectly used truthy checks to determine if the timezone value is present, where `0` is not truthy.

aduth committed:
Components: Refactor date/time components to use configured timezone

aduth committed:
Consolidate "stable" change handling behavior into existing tests

aduth committed:
Align initial date handling for undefined initial time

aduth committed:
Add label for "Today" and ensure alignment with site time

aduth committed:
Add CHANGELOG for fixes

aduth committed:
Fix events number handling for dates

aduth committed:
Update test assertion labels to include new "Today" suffix

aduth committed:
Fix incorrect time handling for UTC internal date converted to site time

aduth committed:
Refactor date utils to use WordPress date API

jasperck reviewed (approved):
Thanks a bunch! I tested from [here](https://github.com/Automattic/dayone-web-client/pull/7), and everything’s looking good now 🎉

aduth committed:
Improve testing assurances with specific non-UTC timezone

aduth committed:
Expand test matrix coverage with "string with timezone" example

This has unique handling in inputToDate (also tested in utils.test.ts), so ensure expected behavior in usage in DateTimePicker and TimePicker components (especially round-trip behavior)

aduth committed:
Ensure Home/End calculates correct start/end of week

aduth committed:
Fix incorrect handling of event dates

Yet another example of fragility in how PostSchedule component initializes date objects using timezoneless strings, but these are not compatible types since date objects reflect the user's current timezone and timezoneless strings are meant to match the site timezone, and there's no guarantees they're the same.

aduth commented:
In testing across this branch and what's in the stable release of WordPress 6.9, one other thing I noticed that this fixes is that the "Today" highlight in 6.9 follows the current browser time rather than the site time. The screenshots below show a scenario where my local time (02:38 2025-12-20UTC+9:30) is a day ahead of the site time (15:59 2025-12-19UTC). My expectation is that this highlight should follow what "Today" is in the site's time (i.e. the 19th), same as how we default to the "current" time in scheduling.

I couldn't find an open issue related to this, but raising for awareness.

State|Before (WordPress 6.9)|After
---|---|---
Initial page load|<img width="346" height="511" alt="image" src="https://github.com/user-attachments/assets/18718319-68de-427f-b8b7-5ba5b5690494" />|<img width="353" height="511" alt="image" src="https://github.com/user-attachments/assets/e9b1ccc7-78cb-40a8-8904-45f9582b10dc" />
Selecting 18th|<img width="353" height="520" alt="image" src="https://github.com/user-attachments/assets/bb4e7f1b-ba0d-4686-beef-07610da33ff6" />|<img width="351" height="515" alt="image" src="https://github.com/user-attachments/assets/4fb36398-5e05-41df-94d6-983ef2c178a2" />

aduth committed:
Add CHANGELOG entry for date package fix
